tokio/src/runtime/builder.rs

```diff
@@ -204,70 +222,129 @@ impl Builder {
+    pub fn clock(&mut self, clock: Clock) -> &mut Self {
+        self.clock = clock;
+        self
+    }
+
     /// Create the configured `Runtime`.
     ///
     /// The returned `ThreadPool` instance is ready to spawn tasks.
     ///
     /// # Examples
     ///
     /// ```
-    /// # use tokio::runtime::Builder;
-    /// # pub fn main() {
-    /// let runtime = Builder::new().build().unwrap();
-    /// // ... call runtime.run(...)
-    /// # let _ = runtime;
-    /// # }
+    /// use tokio::runtime::Builder;
+    ///
+    /// let mut rt = Builder::new().build().unwrap();
+    ///
+    /// rt.block_on(async {
+    ///     println!("Hello from the Tokio runtime");
+    /// });
     /// ```
     pub fn build(&mut self) -> io::Result<Runtime> {
-        let mut reactor_handles = Vec::new();
+        if self.current_thread {
+            self.build_current_thread()
+        } else {
+            self.build_threadpool()
+        }
+    }
+
+    fn build_current_thread(&mut self) -> io::Result<Runtime> {
+        // Create network driver
+        let net = Reactor::new()?;
+        let net_handles = vec![net.handle()];
+
+        let timer = Timer::new_with_clock(net, self.clock.clone());
+        let timer_handles = vec![timer.handle()];
+
+        // And now put a single-threaded executor on top of the timer. When
+        // there are no futures ready to do something, it'll let the timer or
+        // the reactor to generate some new stimuli for the futures to continue
+        // in their life.
+        let executor = CurrentThread::new(timer);
+
+        // Blocking pool
+        let blocking_pool = PoolWaiter::from(Pool::default());
+
+        Ok(Runtime {
+            kind: Kind::CurrentThread(executor),
+            net_handles,
+            timer_handles,
+            blocking_pool,
+        })
+    }
+
+    fn build_threadpool(&mut self) -> io::Result<Runtime> {
+        let mut net_handles = Vec::new();
         let mut timer_handles = Vec::new();
         let mut timers = Vec::new();
 
         for _ in 0..self.num_threads {
-            // Create a new reactor.
-            let reactor = Reactor::new()?;
-            reactor_handles.push(reactor.handle());
+            // Create network driver
+            let net = Reactor::new()?;
+            net_handles.push(net.handle());
 
             // Create a new timer.
-            let timer = Timer::new_with_clock(reactor, self.clock.clone());
+            let timer = Timer::new_with_clock(net, self.clock.clone());
             timer_handles.push(timer.handle());
             timers.push(Mutex::new(Some(timer)));
         }
 
         // Get a handle to the clock for the runtime.
         let clock = self.clock.clone();
 
-        let around_reactor_handles = reactor_handles.clone();
-        let around_timer_handles = timer_handles.clone();
+        // Blocking pool
+        let blocking_pool = PoolWaiter::from(Pool::default());
```

Comment:
This doesn't seem right. The `ThreadPool` builder already creates a blocking pool, and stores a `PoolWaiter` internally. With this you'll end up with two pools.

Reply:
You are correct. This was done intentionally as a transitional step. Fixing this "right" will take some restructuring of `thread_pool` that i want to punt to a follow up PR.

Reply:
Okay, seems fine as long as it'll be fixed before the next release.

Reply:
Maybe leave a `TODO:`?